File: metadata/face-detection.js
const faceapi = require('face-api.js')
const fs = require('fs')
const path = require('path')
const { exec } = require('child_process')
const { Canvas, Image, ImageData, createCanvas, loadImage } = require('canvas')

faceapi.env.monkeyPatch({ Canvas, Image, ImageData })
const outputVideo = path.join('/data/outputs/', 'output.mp4')

async function getFilename() {
  const url =
    'https://raw.githubusercontent.com/oceanprotocol/c2d-examples/main/face-detection/face-demographics-walking-and-pause-short.mp4'
  const response = await fetch(url, { method: 'HEAD' })
  const contentDisposition = response.headers.get('Content-Disposition')

  if (contentDisposition && contentDisposition.includes('filename=')) {
    console.log(
      `file name: ${contentDisposition.split('filename=')[1].replace(/['"]/g, '')}`
    )
    return contentDisposition.split('filename=')[1].replace(/['"]/g, '')
  }
  console.log(`file name: ${url.split('/').pop().split('?')[0]}`)
  // Fallback: Extract from URL if Content-Disposition header is missing
  return url.split('/').pop().split('?')[0]
}

async function getOriginalFrames() {
  const outputDirectoryForOriginalFrames = 'frames'

  if (!fs.existsSync(outputDirectoryForOriginalFrames)) {
    fs.mkdirSync(outputDirectoryForOriginalFrames)
  }
  const inputVideo = await getFilename()

  const ffmpegCommand = `ffmpeg -i '${inputVideo}' '${outputDirectoryForOriginalFrames}/%04d.png'`

  exec(ffmpegCommand, (error) => {
    if (error) {
      console.error('Error:', error)
    } else {
      console.log('Frames extracted successfully.')
    }
  })
}

async function downloadModels(directory) {
  const folderPath = path.join(directory, 'models')
  const urls = [
    'https://github.com/oceanprotocol/c2d-examples/raw/refs/heads/main/face-detection/models/face_expression_model-shard1',
    'https://github.com/oceanprotocol/c2d-examples/raw/refs/heads/main/face-detection/models/face_expression_model-weights_manifest.json',
    'https://github.com/oceanprotocol/c2d-examples/raw/refs/heads/main/face-detection/models/face_landmark_68_model-shard1',
    'https://github.com/oceanprotocol/c2d-examples/raw/refs/heads/main/face-detection/models/face_landmark_68_model-weights_manifest.json',
    'https://github.com/oceanprotocol/c2d-examples/raw/refs/heads/main/face-detection/models/face_recognition_model-shard1',
    'https://github.com/oceanprotocol/c2d-examples/raw/refs/heads/main/face-detection/models/face_recognition_model-shard2',
    'https://github.com/oceanprotocol/c2d-examples/raw/refs/heads/main/face-detection/models/face_recognition_model-weights_manifest.json',
    'https://github.com/oceanprotocol/c2d-examples/raw/refs/heads/main/face-detection/models/tiny_face_detector_model-shard1',
    'https://github.com/oceanprotocol/c2d-examples/raw/refs/heads/main/face-detection/models/tiny_face_detector_model-weights_manifest.json'
  ]
  // Ensure the folder exists
  if (!fs.existsSync(folderPath)) {
    fs.mkdirSync(folderPath, { recursive: true })
  }

  // Function to download a file
  const downloadFile = async (url) => {
    try {
      const response = await fetch(url)
      if (!response.ok) {
        throw new Error(
          `Failed to download ${url}: ${response.status} ${response.statusText}`
        )
      }

      const fileName = path.basename(new URL(url).pathname)
      const filePath = path.join(folderPath, fileName)

      const fileStream = fs.createWriteStream(filePath)
      return new Promise((resolve, reject) => {
        response.body.pipe(fileStream)
        response.body.on('error', reject)
        fileStream.on('finish', () => {
          console.log(`Downloaded: ${fileName}`)
          resolve()
        })
      })
    } catch (error) {
      console.error(`Error downloading ${url}:`, error)
    }
  }

  // Download all files concurrently
  await Promise.all(urls.map(downloadFile))
  console.log('All downloads completed!')
}

async function loadModels(directory) {
  await downloadModels(directory)
  await faceapi.nets.tinyFaceDetector.loadFromDisk('models')
  await faceapi.nets.faceLandmark68Net.loadFromDisk('models')
  await faceapi.nets.faceRecognitionNet.loadFromDisk('models')
  await faceapi.nets.faceExpressionNet.loadFromDisk('models')
}

async function processImages(inputDirectory, outputDirectory) {
  const imageFiles = fs.readdirSync(inputDirectory)

  for (const file of imageFiles) {
    if (file.endsWith('.png')) {
      const inputImagePath = path.join(inputDirectory, file)
      const outputImagePath = path.join(outputDirectory, file)
      const image = await loadImage(inputImagePath)

      const faceDetectionOptions = new faceapi.TinyFaceDetectorOptions()
      const detections = await faceapi
        .detectAllFaces(image, faceDetectionOptions)
        .withFaceLandmarks()
        .withFaceExpressions()

      const canvas = createCanvas(image.width, image.height)
      const ctx = canvas.getContext('2d')
      ctx.drawImage(image, 0, 0)
      faceapi.draw.drawDetections(canvas, detections)

      fs.writeFileSync(outputImagePath, canvas.toBuffer('image/png'))
    }
  }
  console.log('Frame processing complete')
}

function isDirectoryNotEmpty(directoryPath) {
  try {
    const files = fs.readdirSync(directoryPath)

    return files.length > 0
  } catch (error) {
    // Handle errors, such as if the directory doesn't exist
    return false
  }
}

async function main() {
  const inputDirectory = 'frames'
  const outputDirectoryForProcessedFrames = 'processed_frames'

  if (!isDirectoryNotEmpty(inputDirectory)) {
    getOriginalFrames()
  }

  if (!fs.existsSync(outputDirectoryForProcessedFrames)) {
    fs.mkdirSync(outputDirectoryForProcessedFrames)
  }

  await loadModels(inputDirectory)

  await processImages(inputDirectory, outputDirectoryForProcessedFrames)

  console.log('Face detection and processing complete.')

  const ffmpegCommandOutput = `ffmpeg -i ${outputDirectoryForProcessedFrames}/%04d.png -c:v libx264 -pix_fmt yuv420p ${outputVideo}`
  exec(ffmpegCommandOutput, async (error) => {
    if (error) {
      console.error('Error:', error)
      // Try again to process the frames
      if (!fs.existsSync(outputDirectoryForProcessedFrames)) {
        fs.mkdirSync(outputDirectoryForProcessedFrames)
      }
      await processImages(inputDirectory, outputDirectoryForProcessedFrames)
    } else {
      console.log('Frames extracted successfully.')
    }
  })
}

main()

File: metadata/face_detection.py
import cv2
import requests
import os
import subprocess
import glob


# Function to download video from URL
def download_video(url, output_path):
    print("Downloading video...")
    response = requests.get(url, stream=True)
    with open(output_path, 'wb') as file:
        for chunk in response.iter_content(chunk_size=1024):
            if chunk:
                file.write(chunk)
    print("Download complete.")


def extract_frames(video_path, frame_dir):
    """Extract frames from a video using FFmpeg."""
    os.makedirs(frame_dir, exist_ok=True)
    
    command = [
        "ffmpeg", "-i", video_path, os.path.join(frame_dir, "%04d.png")
    ]
    
    subprocess.run(command, check=True)
    print(f"✅ Frames extracted to {frame_dir}")

def detect_faces(image_dir, output_dir):
    """Detects faces in images and saves processed frames."""
    os.makedirs(output_dir, exist_ok=True)

    face_cascade = cv2.CascadeClassifier(cv2.data.haarcascades + "haarcascade_frontalface_default.xml")

    images = sorted(glob.glob(os.path.join(image_dir, "*.png")))
    if not images:
        raise Exception("🚨 No frames found for face detection!")

    for img_path in images:
        img = cv2.imread(img_path)
        gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

        faces = face_cascade.detectMultiScale(gray, scaleFactor=1.3, minNeighbors=5, minSize=(30, 30))

        for (x, y, w, h) in faces:
            cv2.rectangle(img, (x, y), (x + w, y + h), (0, 255, 0), 2)

        output_path = os.path.join(output_dir, os.path.basename(img_path))
        cv2.imwrite(output_path, img)
    
    print(f"✅ Processed {len(images)} frames with face detection.")

    first_frame = cv2.imread(images[0])
    height, width, _ = first_frame.shape

    fourcc = cv2.VideoWriter_fourcc(*'mp4v')
    video_writer = cv2.VideoWriter("/data/outputs/output_faces.mp4", fourcc, 25, (width, height))

    for image in images:
        frame_path = os.path.join(output_dir, os.path.basename(image))
        frame = cv2.imread(frame_path)
        video_writer.write(frame)

    # Release video writer
    video_writer.release()
    print(f"✅ Final video saved at: /data/outputs/output_faces.mp4")

    camera = cv2.VideoCapture("/data/outputs/output_faces.mp4")
    print(camera.isOpened())
    print(camera.read())

    cv2.destroyAllWindows()


if __name__ == "__main__":
    video_url = "https://raw.githubusercontent.com/oceanprotocol/c2d-examples/main/face-detection/face-demographics-walking-and-pause-short.mp4"  # Replace with the actual URL
    video_path = "downloaded_video.mp4"
    frames_dir = "output_frames"
    processed_frames_dir = "processed_frames"
    output_path = "/data/outputs/output_faces.mp4"

    download_video(video_url, video_path)
    extract_frames(video_path, frames_dir)
    detect_faces(frames_dir, processed_frames_dir)

    # Cleanup downloaded video
    os.remove(video_path)